File: chatgpt-app/kst.js
// KST diagnostic engine — JavaScript port of spine/kst.py + spine/diagnose.py,
// for the Node MCP server (Cloud Run can't run the Python engine).
//
// Source of truth is the Python engine + its unit tests (spine/test_kst.py).
// This port is faithful and validated against it (see kst.test.mjs): knowledge
// structure = order ideals of the prerequisite DAG; BLIM slip/guess response
// model; Bayesian posterior over states; outer-fringe gap location; half-split
// next-item. Structure comes ONLY from prereq_edges; calibration ONLY from
// calibration.jsonl — the two KST inputs stay separate.
//
// States are represented as integer bitmasks over the node set (<= ~31 nodes).

import { readFileSync, readdirSync } from "node:fs";
import { join } from "node:path";

function readJsonl(path) {
  const out = [];
  for (const line of readFileSync(path, "utf8").split("\n")) {
    const s = line.trim();
    if (!s || s.startsWith("#")) continue;
    out.push(JSON.parse(s));
  }
  return out;
}

// ---- structure: prerequisite DAG -> knowledge states (order ideals) ----
function buildStructure(edges) {
  const order = [];
  const seen = new Set();
  for (const e of edges) {
    if ((e.type || "prerequisite") !== "prerequisite") continue;
    for (const x of [e.from, e.to]) if (!seen.has(x)) { seen.add(x); order.push(x); }
  }
  const idx = new Map(order.map((n, i) => [n, i]));
  const n = order.length;
  const predMask = new Array(n).fill(0);
  for (const e of edges) {
    if ((e.type || "prerequisite") !== "prerequisite") continue;
    predMask[idx.get(e.to)] |= 1 << idx.get(e.from);
  }
  // children adjacency (successors)
  const children = Array.from({ length: n }, () => []);
  for (let j = 0; j < n; j++)
    for (let i = 0; i < n; i++)
      if ((predMask[j] >> i) & 1) children[i].push(j);
  // topological depth
  const depth = new Array(n).fill(-1);
  const dfsDepth = (i) => {
    if (depth[i] >= 0) return depth[i];
    let d = 0;
    for (let p = 0; p < n; p++) if ((predMask[i] >> p) & 1) d = Math.max(d, 1 + dfsDepth(p));
    return (depth[i] = d);
  };
  for (let i = 0; i < n; i++) dfsDepth(i);

  // enumerate ideals via BFS: add a node iff all its prereqs are present
  const states = new Set([0]);
  const frontier = [0];
  while (frontier.length) {
    const k = frontier.pop();
    for (let i = 0; i < n; i++) {
      if ((k >> i) & 1) continue;
      if ((predMask[i] & k) === predMask[i]) {
        const k2 = k | (1 << i);
        if (!states.has(k2)) { states.add(k2); frontier.push(k2); }
      }
    }
  }
  return { order, idx, n, predMask, children, depth, states: [...states] };
}

function reaches(struct, src, targetSet) {
  if (targetSet.has(src)) return true;
  const seen = new Set([src]);
  const stack = [...struct.children[src]];
  while (stack.length) {
    const x = stack.pop();
    if (targetSet.has(x)) return true;
    if (!seen.has(x)) { seen.add(x); for (const c of struct.children[x]) stack.push(c); }
  }
  return false;
}

// ---- BLIM + Bayesian assessment ----
const pCorrect = (inState, slip, guess) => (inState ? 1 - slip : guess);

function posterior(struct, responses, items, calib) {
  const states = struct.states;
  let post = new Map(states.map((s) => [s, 1 / states.length]));
  for (const [itemId, correct] of responses) {
    const node = items.get(itemId);
    if (node == null) continue;
    const i = struct.idx.get(node);
    const c = calib.get(itemId) || { slip: 0.1, guess: 0.1 };
    const next = new Map();
    let Z = 0;
    for (const [k, w] of post) {
      const inState = i !== undefined && ((k >> i) & 1) === 1;
      const p = pCorrect(inState, c.slip, c.guess);
      const lik = correct ? p : 1 - p;
      const v = w * lik;
      next.set(k, v); Z += v;
    }
    if (Z <= 0) continue;
    for (const [k, v] of next) next.set(k, v / Z);
    post = next;
  }
  return post;
}

function predictedPCorrect(struct, post, itemId, items, calib) {
  const node = items.get(itemId);
  if (node == null) return null;
  const i = struct.idx.get(node);
  const c = calib.get(itemId) || { slip: 0.1, guess: 0.1 };
  let s = 0;
  for (const [k, w] of post) {
    const inState = i !== undefined && ((k >> i) & 1) === 1;
    s += w * pCorrect(inState, c.slip, c.guess);
  }
  return s;
}

function mapState(post) {
  let best = 0, bestKey = [-1, 0];
  for (const [k, w] of post) {
    const key = [w, -popcount(k)];
    if (key[0] > bestKey[0] || (key[0] === bestKey[0] && key[1] > bestKey[1])) { best = k; bestKey = key; }
  }
  return best;
}
const popcount = (x) => { let c = 0; while (x) { x &= x - 1; c++; } return c; };

function outerFringe(struct, mask) {
  const out = [];
  for (let i = 0; i < struct.n; i++)
    if (!((mask >> i) & 1) && (struct.predMask[i] & mask) === struct.predMask[i]) out.push(i);
  return out;
}

function locateGap(struct, responses, items, calib) {
  const post = posterior(struct, responses, items, calib);
  const K = mapState(post);
  const fringe = outerFringe(struct, K);
  const failed = new Set();
  for (const [itemId, correct] of responses) {
    if (correct) continue;
    const node = items.get(itemId);
    if (node != null && struct.idx.has(node)) failed.add(struct.idx.get(node));
  }
  const rank = (i) => [failed.size && reaches(struct, i, failed) ? 0 : 1, struct.depth[i], i];
  let primary = null;
  for (const i of fringe) {
    if (primary === null) { primary = i; continue; }
    const a = rank(i), b = rank(primary);
    if (a[0] < b[0] || (a[0] === b[0] && (a[1] < b[1] || (a[1] === b[1] && a[2] < b[2])))) primary = i;
  }
  return { post, K, fringe, primary, failed: [...failed], confidence: post.get(K) || 0 };
}

function nextItemIdx(struct, responses, items, calib, asked) {
  const post = posterior(struct, responses, items, calib);
  const seen = new Set(asked);
  for (const [id] of responses) seen.add(id);
  let best = null, bestGap = 2;
  for (const [itemId, node] of items) {
    if (node == null || seen.has(itemId)) continue;
    const p = predictedPCorrect(struct, post, itemId, items, calib);
    const gap = Math.abs(0.5 - p);
    if (gap < bestGap) { best = itemId; bestGap = gap; }
  }
  return best;
}

// ---- diagnostics facade (mirrors spine/diagnose.py) ----
export function createDiagnostics({ spineDir }) {
  const edges = readJsonl(join(spineDir, "prereq_edges.jsonl"));
  const struct = buildStructure(edges);
  // slug -> {concept_en, lesson, book_pages}, the 26 curriculum concepts
  // (from INDEX.json; library.json only carries the 6 video concepts).
  const concepts = JSON.parse(readFileSync(join(spineDir, "concepts.json"), "utf8"));

  const itemRows = readJsonl(join(spineDir, "items.jsonl"));
  const items = new Map(itemRows.map((r) => [r.item_id, r.node ?? null]));
  const itemMeta = new Map(itemRows.map((r) => [r.item_id, r]));
  const calib = new Map(
    readJsonl(join(spineDir, "calibration.jsonl")).map((r) => [r.item_id, { slip: r.slip, guess: r.guess }])
  );

  const crosswalk = readJsonl(join(spineDir, "crosswalk.jsonl"));
  const nodeToSlugs = new Map();
  const label = new Map();
  for (const r of crosswalk) {
    if (r.match !== "none" && r.node) {
      if (!nodeToSlugs.has(r.node)) nodeToSlugs.set(r.node, []);
      nodeToSlugs.get(r.node).push(r.slug);
      if (r.label) label.set(r.node, r.label);
    }
  }
  const teachableIdx = new Set(
    [...nodeToSlugs.keys()].filter((nd) => struct.idx.has(nd)).map((nd) => struct.idx.get(nd))
  );
  const nodeOf = (i) => struct.order[i];
  const labelOf = (nd) => label.get(nd) || nd;
  const conceptInfo = (slug) => {
    const c = concepts[slug] || {};
    return { slug, concept_en: c.concept_en, lesson: c.lesson, book_pages: c.book_pages };
  };

  function teachTarget(nodeIdx) {
    const nd = nodeOf(nodeIdx);
    const slugs = nodeToSlugs.get(nd);
    if (slugs && slugs.length) return { kind: "concept", payload: slugs.map(conceptInfo) };
    // floor node: nearest downstream teachable concept
    const seen = new Set([nodeIdx]);
    const q = [...struct.children[nodeIdx]];
    while (q.length) {
      const j = q.shift();
      if (seen.has(j)) continue;
      seen.add(j);
      const s = nodeToSlugs.get(nodeOf(j));
      if (s && s.length) return { kind: "prerequisite", payload: { node: nd, label: labelOf(nd), motivates: conceptInfo(s[0]) } };
      for (const c of struct.children[j]) q.push(c);
    }
    return { kind: "prerequisite", payload: { node: nd, label: labelOf(nd), motivates: null } };
  }

  function teachingTarget(g) {
    const mastered = g.K;
    const failed = new Set(g.failed);
    let best = null;
    for (const i of teachableIdx) {
      if ((mastered >> i) & 1) continue;
      if (failed.has(i) || (failed.size && reaches(struct, i, failed))) {
        if (best === null || struct.depth[i] < struct.depth[best] || (struct.depth[i] === struct.depth[best] && i < best)) best = i;
      }
    }
    return best !== null ? best : g.primary;
  }

  function diagnose(responses) {
    const resp = responses.map((r) => [r.item_id, !!r.correct]);
    const g = locateGap(struct, resp, items, calib);
    const used = resp.filter(([id]) => items.get(id) != null).length;
    const result = {
      n_states: struct.states.length,
      confidence: Math.round(g.confidence * 1e4) / 1e4,
      responses_used: used,
      responses_ignored: resp.length - used,
      structural_gap: g.primary != null ? { node: nodeOf(g.primary), label: labelOf(nodeOf(g.primary)) } : null,
      failed_nodes: g.failed.map((i) => ({ node: nodeOf(i), label: labelOf(nodeOf(i)) })),
      fringe: g.fringe.map((i) => ({ node: nodeOf(i), label: labelOf(nodeOf(i)) })),
    };
    const target = teachingTarget(g);
    if (target == null) {
      return { ...result, status: "no_gap", gap_node: null,
        message: "No gap located: the student's mastery spans the assessed spine." };
    }
    const tt = teachTarget(target);
    result.gap_node = nodeOf(target);
    result.gap_label = labelOf(nodeOf(target));
    result.status = "gap_located";
    result.teach_kind = tt.kind;
    result.teach = tt.payload;
    if (tt.kind === "concept") {
      const primary = tt.payload[0];
      Object.assign(result, { slug: primary.slug, lesson: primary.lesson,
        book_pages: primary.book_pages, concept_en: primary.concept_en });
      if (tt.payload.length > 1) result.note = `teaching node maps to multiple collapsed concepts: ${tt.payload.map((p) => p.slug).join(", ")}`;
    } else {
      result.note = "gap is a prerequisite/math-floor node with no Physics-2 lesson; teach the prerequisite, motivated by the concept it unlocks.";
    }
    const structuralNode = result.structural_gap && result.structural_gap.node;
    if (structuralNode && structuralNode !== result.gap_node) {
      result.note = ((result.note || "") +
        ` KST structural gap is deeper (${result.structural_gap.label}); intermediate nodes carry no items, so the teachable target is reported. Low confidence reflects sparse item coverage.`).trim();
    }
    return result;
  }

  function suggestNext(responses) {
    const resp = responses.map((r) => [r.item_id, !!r.correct]);
    const id = nextItemIdx(struct, resp, items, calib, []);
    if (!id) return { next_item: null };
    const m = itemMeta.get(id) || {};
    return { next_item: { item_id: id, lesson: m.lesson, node: m.node, qkind: m.qkind, printed_number: m.printed_number } };
  }

  return { diagnose, suggestNext, _struct: struct, _items: items, _calib: calib };
}

File: spine/kst.py
#!/usr/bin/env python3
"""Knowledge Space Theory engine for the Physics-2 diagnostic spine (Phase 5).

Pure functions, stdlib only. Two inputs kept strictly separate, exactly as KST
requires:

  * STRUCTURE  — which knowledge states are possible. Derived ONLY from the
                 prerequisite DAG (the graph). See `knowledge_states`.
  * CALIBRATION — per-item slip/guess. Derived ONLY from response data (the
                 dataset / defaults). Passed in as a dict, never inferred here.

The engine combines structure + calibration + a student's answers to locate the
gap. It is the deterministic authority; the LLM never decides the structure.

Algorithms (cited to Doignon & Falmagne, *Knowledge Spaces* / *Learning Spaces*):
  * knowledge_states  -> the knowledge structure as the family of order ideals
                         (downward-closed sets) of the prerequisite order. For a
                         partial order this family is a quasi-ordinal knowledge
                         space, closed under union and intersection (Birkhoff's
                         representation; D&F ch. 3).
  * p_correct         -> Basic Local Independence Model (BLIM): a single item's
                         response depends only on whether its node is in the
                         state, via that item's slip (beta) and guess (eta).
                         (D&F ch. 11 / Falmagne-Doignon BLIM.)
  * posterior         -> Bayesian update of the likelihood over states given
                         observed responses (the stochastic assessment kernel,
                         D&F ch. 13 / ch. 10 questioning).
  * next_item         -> half-split questioning rule: ask the item whose
                         predicted P(correct) under the current belief is closest
                         to 1/2 (maximally informative single question).
  * locate_gap        -> the outer fringe of the most likely state: the nodes a
                         learner is "ready to learn" (all prerequisites mastered,
                         node itself not). The fringe is the teachable gap.
"""
from __future__ import annotations

import math
from collections import defaultdict
from dataclasses import dataclass


# ----------------------------------------------------------------------------
# Structure: prerequisite order -> knowledge states (order ideals)
# ----------------------------------------------------------------------------

@dataclass(frozen=True)
class Structure:
    nodes: tuple                      # all node ids
    preds: dict                       # node -> frozenset(direct prerequisites)
    succ: dict                        # node -> frozenset(direct successors)
    depth: dict                       # node -> topological depth (roots = 0)
    states: tuple                     # tuple of frozensets, the knowledge states

    def reaches(self, src, targets):
        """True if any node in `targets` is reachable from src via successors."""
        if src in targets:
            return True
        seen, stack = {src}, list(self.succ.get(src, ()))
        while stack:
            n = stack.pop()
            if n in targets:
                return True
            if n not in seen:
                seen.add(n)
                stack.extend(self.succ.get(n, ()))
        return False


def _direct_preds(nodes, edges):
    preds = {n: set() for n in nodes}
    for e in edges:
        if e.get("type", "prerequisite") != "prerequisite":
            continue
        preds[e["to"]].add(e["from"])
    return {n: frozenset(p) for n, p in preds.items()}


def _depths(nodes, preds):
    depth, seen = {}, {}

    def d(n, stack=()):
        if n in depth:
            return depth[n]
        if not preds[n]:
            depth[n] = 0
            return 0
        depth[n] = 1 + max(d(p) for p in preds[n])
        return depth[n]

    for n in nodes:
        d(n)
    return depth


def knowledge_states(edges, nodes=None, max_states=200_000):
    """Enumerate the knowledge structure: all downward-closed sets of the
    prerequisite order. A set K is a state iff every node in K has all of its
    prerequisites in K (you cannot know a thing without its prerequisites).

    Raises ValueError if the count would exceed `max_states` (caller should then
    restrict the structure to a relevant sub-order; see diagnose.py).
    """
    if nodes is None:
        nodes = []
        seen = set()
        for e in edges:
            for x in (e["from"], e["to"]):
                if x not in seen:
                    seen.add(x)
                    nodes.append(x)
    nodes = tuple(nodes)
    preds = _direct_preds(nodes, edges)

    # BFS over ideals: start from the empty state, add any node all of whose
    # prerequisites are already present. Every ideal is reachable this way (add
    # elements along any linear extension).
    empty = frozenset()
    states = {empty}
    frontier = [empty]
    while frontier:
        K = frontier.pop()
        for n in nodes:
            if n in K:
                continue
            if preds[n] <= K:
                K2 = K | {n}
                if K2 not in states:
                    states.add(K2)
                    frontier.append(K2)
                    if len(states) > max_states:
                        raise ValueError(
                            f"knowledge structure exceeds {max_states} states; "
                            "restrict to a sub-order before enumerating."
                        )
    depth = _depths(nodes, preds)
    succ = {n: set() for n in nodes}
    for n in nodes:
        for p in preds[n]:
            succ[p].add(n)
    succ = {n: frozenset(s) for n, s in succ.items()}
    # Stable ordering: by size then by sorted contents.
    ordered = tuple(sorted(states, key=lambda s: (len(s), sorted(s))))
    return Structure(nodes=nodes, preds=preds, succ=succ, depth=depth, states=ordered)


# ----------------------------------------------------------------------------
# Response model (BLIM) + Bayesian assessment
# ----------------------------------------------------------------------------

def p_correct(node_in_state: bool, slip: float, guess: float) -> float:
    """BLIM: P(correct | state). Knows the node -> 1-slip; doesn't -> guess."""
    return (1.0 - slip) if node_in_state else guess


def posterior(structure: Structure, responses, items, calib, prior=None):
    """Bayesian posterior over knowledge states given observed responses.

    responses : iterable of (item_id, correct: bool)
    items     : item_id -> node id (the node the item assesses; None items skipped)
    calib     : item_id -> {"slip":.., "guess":..}
    prior     : optional state -> weight; defaults to uniform.
    Returns: dict state(frozenset) -> normalized probability.
    """
    states = structure.states
    if prior is None:
        post = {K: 1.0 / len(states) for K in states}
    else:
        Z = sum(prior.values())
        post = {K: prior.get(K, 0.0) / Z for K in states}

    for item_id, correct in responses:
        node = items.get(item_id)
        if node is None:
            continue  # unresolved/ambiguous item carries no structural signal
        c = calib.get(item_id, {"slip": 0.10, "guess": 0.10})
        slip, guess = c["slip"], c["guess"]
        new = {}
        for K, w in post.items():
            if w == 0.0:
                new[K] = 0.0
                continue
            p = p_correct(node in K, slip, guess)
            lik = p if correct else (1.0 - p)
            new[K] = w * lik
        Z = sum(new.values())
        if Z <= 0:
            # Degenerate (impossible under model); keep prior rather than divide by 0.
            continue
        post = {K: w / Z for K, w in new.items()}
    return post


def predicted_p_correct(structure, post, item_id, items, calib):
    """Belief-weighted P(correct) for an item under the current posterior."""
    node = items.get(item_id)
    if node is None:
        return None
    c = calib.get(item_id, {"slip": 0.10, "guess": 0.10})
    return sum(w * p_correct(node in K, c["slip"], c["guess"])
               for K, w in post.items())


def next_item(structure, responses, items, calib, asked=()):
    """Half-split rule: among unasked, resolvable items, pick the one whose
    predicted P(correct) is closest to 0.5 (the most informative question)."""
    post = posterior(structure, responses, items, calib)
    asked = set(asked) | {r[0] for r in responses}
    best, best_gap = None, 2.0
    for item_id, node in items.items():
        if node is None or item_id in asked:
            continue
        p = predicted_p_correct(structure, post, item_id, items, calib)
        gap = abs(0.5 - p)
        if gap < best_gap:
            best, best_gap = item_id, gap
    return best


def map_state(structure, post):
    """Maximum-a-posteriori knowledge state."""
    return max(post.items(), key=lambda kv: (kv[1], -len(kv[0])))[0]


def outer_fringe(structure, state):
    """Nodes the learner is ready to learn: not yet known, all prereqs known."""
    return [n for n in structure.nodes
            if n not in state and structure.preds[n] <= state]


def marginal_mastery(structure, post):
    """Per-node P(node is in the student's state)."""
    m = defaultdict(float)
    for K, w in post.items():
        for n in K:
            m[n] += w
    return {n: m.get(n, 0.0) for n in structure.nodes}


def locate_gap(structure, responses, items, calib, confident=0.0):
    """Locate the diagnostic gap.

    Returns dict with the MAP state, its outer fringe, the primary gap (most
    foundational fringe node = lowest topological depth), per-node mastery, and a
    confidence = posterior mass on the MAP state.
    """
    post = posterior(structure, responses, items, calib)
    K = map_state(structure, post)
    fringe = outer_fringe(structure, K)

    # Nodes the student demonstrably failed (an item on them answered wrong).
    failed = frozenset(structure_node for (item_id, correct) in responses
                       if not correct and (structure_node := items.get(item_id)))

    # Primary gap: the most foundational unmet prerequisite ON THE PATH to a
    # concept the student actually failed. This avoids sending a student who
    # missed momentum to an unrelated root (e.g. Potential energy) when the
    # posterior is diffuse because intermediate nodes carry no items. If nothing
    # failed (or no fringe node leads to a failure), fall back to global depth.
    def rank(n):
        return (0 if (failed and structure.reaches(n, failed)) else 1,
                structure.depth[n], n)
    primary = min(fringe, key=rank) if fringe else None

    return {
        "map_state": K,
        "fringe": fringe,
        "primary_gap": primary,
        "confidence": post.get(K, 0.0),
        "failed_nodes": sorted(failed),
        "mastery": marginal_mastery(structure, post),
        "n_states": len(structure.states),
    }
